### ISSUE TITLE
fs::rename does not delete the source file if it's a hard-link of the dest file (linux)

### ISSUE THREAD

**Issue by lefth** [rust-lang/rust#91326]
When performing move/rename, the source file should no longer exist after the operation, regardless of its relationship with the destination file (unless they are the same file). I found that this does not happen if the two files are hard linked. I think the sameness check (_are src and dest the same file?_) may have a false positive in this case. I guess the problem is that it's not always trivial to check whether two paths refer to the same file, but comparing inodes for equality doesn't give the right answer.

Here is a simple program that should link "a.txt" to "b.txt", overwriting if necessary:

```rust
fn main() -> std::io::Result<()> {
    std::fs::hard_link("a.txt", "temp_file")?;
    std::fs::rename("temp_file", "b.txt")?;
    Ok(())
}
```

The first time it runs, it operates as expected, with `temp_file` no longer existing because it has been renamed:
```shell
$ touch a.txt b.txt
$ cargo run --quiet && ls
a.txt  b.txt  Cargo.lock  Cargo.toml  src  target
```

The second time it runs, `temp_file` is not deleted as part of the rename operation:
```shell
$ cargo run --quiet && ls
a.txt  b.txt  Cargo.lock  Cargo.toml  src  target  temp_file
```

The source argument should be deleted when a rename is successful. Because although the semantics of this case are weird, a user should always be able to assume that if `fs::rename` is successful (and if there is no race condition), the source path no longer exists, and the destination path contains the file that used to be at the source path.

### Meta
`rustc --version --verbose`:
```
rustc 1.57.0-nightly (2c7bc5e33 2021-09-15)
binary: rustc
commit-hash: 2c7bc5e33c25e29058cbafefe680da8d5e9220e9
commit-date: 2021-09-15
host: x86_64-unknown-linux-gnu
release: 1.57.0-nightly
LLVM version: 13.0.0
```
The issue also occurs on stable 1.55.0.

Note: the environment is Linux (WSL2). The filesystem is the virtual Linux filesystem, not a native Windows path. Please let me know if there's any other information I can provide.

**Comment by FabianWolff** [comment]
But the [documentation](https://doc.rust-lang.org/std/fs/fn.rename.html) says that the behavior of `std::fs::rename()` is platform-specific, and that
> This function currently corresponds to the `rename` function on Unix

And the [man page](https://man7.org/linux/man-pages/man2/rename.2.html) for `rename(2)` says that
> If _oldpath_ and _newpath_ are existing hard links referring to the same file, then `rename()` does nothing, and returns a success status.

So it seems that this is the intended behavior.

**Comment by hkratz** [comment]
Interestingly enough Macos behaves the same, though it is not documented. It is also not documented as a special case [in the POSIX standard](https://pubs.opengroup.org/onlinepubs/9699919799/functions/rename.html) or for BSD. I think just calling into libc is the right choice here. Doing any magic ourselves will likely break user code in unexpected ways.

**Comment by FabianWolff** [comment]
> It is also not documented as a special case [in the POSIX standard](https://pubs.opengroup.org/onlinepubs/9699919799/functions/rename.html)

Actually, I think it is, in the third paragraph of the description:
> If the _old_ argument and the _new_ argument resolve to [...] different directory entries for [i.e., hard links to] the same existing file, `rename()` shall return successfully and perform no other action.

**Comment by hkratz** [comment]
Yeah, missed that because I thought that was only about symbolic links.

**Comment by lefth** [comment]
Thank you for explaining the rationale. I will work around this in my program, just as many other programs with "rename" functionality must be doing.
